### ISSUE TITLE
no option for zero-based graphs

### ISSUE THREAD

**Issue by robey** [prometheus/prometheus#520]
I thought there was consensus that graphs should start at 0 on the Y axis to avoid misinterpretation, but a quick google search shows that there's not total agreement on that.

So, I would like the option to make the graph zero-based, without forcing everyone else to agree. :)

For some stats (like "users online"), the graph may stay flat for hours at a time. Currently, prometheus will draw such a graph as blank, with a minimal Y range, and a single line at the baseline where it's effectively invisible.

**Comment by juliusv** [comment]
Agreed. I believe 0-based is the default in PromDash, though it's possible to manually set a Y-axis range (but no option to simply dynamically clamp the Y-axis to the available values).

0-based Y-axes have the problem that very large counters will just look like flat lines even if they're increasing rapidly. But yeah, the reverse problems also exist.

So I think this might actually be worth another UI button, both in Prometheus and in PromDash.

**Comment by brian-brazil** [comment]
Note that http://prometheus.io/docs/visualization/consoles/#graph-library has a `min` option.

**Comment by juliusv** [comment]
Yeah, it's less about the graphing library support than about the question of what UI elements to add to get a balance between UI clutter and really needed features (for the server-side built-in graphing at lesat).

**Comment by brian-brazil** [comment]
I think for /graph we should do something a bit more reasonable when all the values are the same, such as say making the main and max 10% either side of the value.

**Comment by discordianfish** [comment]
I just realize how problematic this can be: "Peaks in the metrics are lower than they appear" :)
Do we have consensus how it should look like? Making it 10% will help with the invisible steady metric but won't change that it's sometime visually misleading. I guess a simply checkbox with whatever default would make most sense since I believe there are scenarios for both.

**Comment by matthiasr** [comment]
another case: stacked graphs should _always_ include the zero line. Right now I'm looking at a stacking of 5 `up == 1` and it's … weird.

![prometheus_time_series_collection_and_processing_server](https://cloud.githubusercontent.com/assets/131590/20519084/61d1d9ac-b099-11e6-8db0-efa2786891c2.png)

**Comment by theophani** [comment]
@matthiasr A stacked graph was exactly what I was looking at when we discussed this yesterday in our company channel. Maybe that is a nice mini-change.

In this graph, I was trying to understand the contribution that the green area contributed to the whole. For a moment, I was convinced it contributed more than half.

![screen shot 2016-11-22 at 11 02 03](https://cloud.githubusercontent.com/assets/364246/20519403/228e9618-b0a3-11e6-8475-b0daef080a85.png)

I understand the desire to see “the interesting” part of the graph in detail, but I think that should be an active choice, not the other way around.

**Comment by discordianfish** [comment]
This seems easy to fix. Are we in agreement that it should be changed? And how is it suppose to look? Just another button beside `[ ] Stacked`? I'd also say we make it default for all graphs.

**Comment by matthiasr** [comment]
👍

**Comment by SameerMahajan-GSLab** [comment]
Is this issue still Open as shown on top or now Closed as marked towards the end of the discussion?

**Comment by kg-mpg** [comment]
Happy to look into this if no one else is

**Comment by beorn7** [comment]
This should now be addressed in the new UI (coming with v3).

**Comment by juliusv** [comment]
> This should now be addressed in the new UI (coming with v3).

Uhm. That'd be news to me?

**Comment by juliusv** [comment]
I see, yeah. I don't know if I'd want to add it to the new UI to be honest. There's already again more buttons and options and clutter than I wanted, and we should draw the line somewhere...

**Comment by beorn7** [comment]
OK, given that nobody worked on this issue in more than three years (despite tags like "help wanted" and "low hanging fruit"), maybe it isn't even desired that much.

I will simply close this issue. If people feel it is really needed _and_ if they have concrete plans to implement it (and of course are willing to have the discussion if it is a net gain for UX), please follow up here, and we can re-open as needed.

**Comment by kokes** [comment]
I have actually wanted this for years now and I'd usually default to stacked charts and isolating a single series, which is the only way to achieve this, currently.

I took a stab at this and produced a POC, linked above. There's also [this deployment preview](https://deploy-preview-14887--prometheus-react.netlify.app/query?g0.expr=container_memory_rss%7Bid%3D%22%2Fsystem.slice%2Fgrafana-server.service%22%7D&g0.show_tree=0&g0.tab=graph&g0.range_input=1h&g0.anchor_y_axis=1&g0.res_type=auto&g0.res_density=medium&g0.display_mode=lines&g0.show_exemplars=0), where you can try the checkbox in action.

(I won't comment on the UI clutter implications, I'm not enough of an insider to have a strong opinion on that.)

**Comment by beorn7** [comment]
I don't have a strong opinion of my own, and I'm not a UI expert anyway.

@juliusv I leave it to you to decide if you want to re-open this, based on the PoC linked above.

**Comment by juliusv** [comment]
So that's a bit of a philosophical question of which bells and whistles we want to add to Prometheus' own UI, and when to send people to Grafana instead. I can see the argument that this option is actually helpful not only for dashboarding, but when just digging around during an incident and wanting to interpret the data better (which is pretty much the use case we want to cover with the Prometheus UI, not full dashboarding).

Currently I already know of at least one other checkbox that I'll need to add (but already styled it differently, as a button), and with that one I'm already running into the situation that the graph options don't fit into a single line anymore on narrower windows. Maybe a decent-enough solution to all this would be to move some of the visualization settings out of the way into a dropdown submenu, just like I did with the "..." menu on the expression input.

We have the big Prometheus 3.0 release tomorrow at PromCon, so not sure if I'll get to play with that idea before then, but it sounds worth at least exploring, so reopening for now.

**Comment by adity-a34** [comment]
Hi @juliusv, 
My research shows no anchor-to-zero implementation exists in the current codebase, 
And the draft PR #14887 is stalled. 
Given the request's history and your prior input on UI concerns, could you clarify if there’s a preferred UX direction or next step to move this feature forward?

I would like to get this issue assigned and propose a PR that optimally resolves it.
Thank you.

**Comment by adity-a34** [comment]
> You previously noted that zero-based graphing is valuable for incident use cases, but adding such options raises concerns over UI clutter.

> You suggested moving extra visualization settings to a dropdown for better UX, though pending the 3.0 release.

Given this, could you specify the desired approach for:

1. Placement and UI style of a zero-anchor toggle (e.g., checkbox, dropdown submenu, etc.)?
2. Any constraints or standards for integrating new graph options?
3. Additional guidelines to align the implementation with your UX vision for Prometheus?

**Comment by juliusv** [comment]
@ADITYATIWARI342005 Thanks, I actually already have a local change that sets the Y axis to zero, the only reason I never got around to finishing it was figuring out the exact aesthetics of adding too much settings clutter to the UI. Let me try and get back to that the next days...

**Comment by juliusv** [comment]
I've got a PR now at: https://github.com/prometheus/prometheus/pull/17565
